### ISSUE TITLE
[explicit-literal-value] new rule for nominal typing (require type annotations)

### ISSUE THREAD

**Issue by Aetet** [typescript-eslint/typescript-eslint#1467]
I'd like to propose new rule.

Description
Typing with array and object literals leads to loose typing. And make it hard to find all references to interface at project. Because literals has different types and no warning for this:

```ts
interface User {
  name: string
}

function validateUser(user: User) {}

const user = {name: 'John'}
validateUser(user);
```

If I want to find all references to User interface:
1. Find validateUser.
2. Find all usages validateUser.
3. Find objects passed to validateUser.

But if I rewrote code to
```ts
const user: User = {name: 'John'}
validateUser(user);
```

it requires from me only step 1.

My propose enforce object literal creation with types:
```ts
// Error - lacks typing, because literal type object has no explicit types
const a = {name: 'John'};
// Error - lacks typing for array and object inside it
const b = [{a: 1}]

// This warns if literalsAsArgument: false
validateUser({name: 'John'})

// This passes because const treat this as constant value
const c = {name: 'John'} as const

// This passes because of casting d to User
const d = {} as User
d.name = 'Jackie boy'

// This passes because of explicit type
const e: User = {name: 'Jill'}
```

Structural typing on the wild leads to error prone refactoring and usually this mistakes discovered on production, but not at compile time.

**Comment by bradzacher** [comment]
You can achieve this to a great deal with the [`typedef`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/typedef.md) rule with the `variableDeclaration` option on.

It will force you to annotate every variable, but it's a step in the direction you want.

Happy to accept a PR there to enhance the flag so that it accepts a string literal. i.e.
```ts
variableDeclaration?: boolean | {
  default?: boolean; // true if you want to require typedefs by default
  objectLiteral?: boolean; // override the default for object literals
  arrayLiteral?: boolean; // override the default for array literals
}
```

**Comment by Zamiell** [comment]
Welcome to 2023. In today's modern era, we have the `satisfies` operator.

So for those who are reading this thread now, it's worth nothing that:

```ts
const user: User = { name: 'John' };
```

Should probably be written instead as:

```ts
const user = { name: 'John' } as const satisfies User;
```

Subsequently, I think that the hypothetical ESLint rule should require the `satisfies` operator, not the type annotation.

**Comment by bradzacher** [comment]
> I think that the hypothetical ESLint rule should require the satisfies operator, not the type annotation as the OP first requested

For the most part they're the same thing.
It's really only cases with indexers (eg `Record<string, value>`) where an annotation and `satisfies` differ.

Regardless - the distinction between the two doesn't matter for the rule.
In essence the rule would only want to know about whether or not the object has a "contextual type" or not (see some of the discussion on the WIP PR #5666).

Both an annotation or `satisfies` will both provide a contextual type - so the choice between them is a stylistic decision outside the scope of the rule!

**Comment by Josh-Cena** [comment]
For `satisfies` vs. type annotation, see also https://github.com/typescript-eslint/typescript-eslint/issues/6347

**Comment by OliverJAsh** [comment]
I open sourced the lint rule we're using at Unsplash: https://github.com/unsplash/eslint-plugin-require-object-type-annotations

More info here: https://oliverjash.me/2023/prefer-explicit-type-annotations-for-objects

**Comment by Josh-Cena** [comment]
@bradzacher I remember you saying `typedef` is only meant as a migration tool. I personally find it hard to configure too, so I just turn it off despite valid cases like this. Shouldn't we just build another rule instead of overloading this rule (which isn't type-aware anyway)?

**Comment by bradzacher** [comment]
This issue has been opened for 3 years and has gotten just 5 reactions.
Unsplash has graciously OSS'd their solution and their plugin currently have 0 weekly downloads.

Honestly at this point I'm inclined to close this feature as something most of the community isn't interested in. To be clear - I see some value in this idea from a maintenance perspective - but there's no point us building/maintaining a rule if nobody's going to turn it on.

---

@Josh-Cena yeah it is - 3 years ago I was looking at incremental changes we could make rather than whole new rules.

**Comment by Josh-Cena** [comment]
I would personally find value in this, but I see how it is hard to build (we need to ask the type checker if the variable has a fresh literal type or a known type inferred from elsewhere), and can lead to lots of false positives. In general I think it isn't worth the effort either just for DX improvements (no correctness improvement).

**Comment by bradzacher** [comment]
Yeah I agree - it feels like adding a DevX burden (extra annotations and lint errors) in exchange for a DevX win (easier refactors). I'm not 100% convinced this is a "net positive" tradeoff personally - but I can see how people might feel otherwise.

Yeah given that there is an OSS solution for this niche usecase - I think it's best to close this.

If the plugin does find some usage amongst the community we can revisit bringing into the fold - but for now we'll leave it 3rd party.
